grafana/grafana
# Explore: Add feature to open log sample in split view

**What is this feature?**

In this PR, we are adding a new feature to Explore's logs sample implemented in https://github.com/grafana/grafana/pull/61864. It is a button to open logs samples in split view. After users clicks on the button, log query that was executed is open in split view.

Known limitations:
1. It will only open the first log query in the split view. This should cover the main usecase, but in a rare cases when user run 2+ loki metrics queries, split view will open just the first one. This is due to the limitation of `splitOpen` API in `@grafana/data` which allows to pass only 1 query. I have created an issue for this https://github.com/grafana/grafana/issues/62107.
2. Mixed data sources behaviour is buggy (in the same way as log context, log filtering, log histogram,...), but I am waiting for https://github.com/grafana/grafana/issues/56314 to be resolved to know how to approach this (added logs sample) to list. 

https://user-images.githubusercontent.com/30407135/214561206-4fd8634a-842c-405e-a033-d05e9d4d5e86.mov

**Why do we need this feature?**

This allows users to jump from limited log samples functionality to running full query and having posibility to modify it/interact with logs. 

**Special notes for your reviewer**:

To be able to get correct log query for `Open logs in split view`, I added `getSupplementaryQuery` to `DataSourceWithSupplementaryQueriesSupport` API. This means that data sources have to have a method that turns original query into supplementary query. 

I then reused this method in `getLogsVolumeDataProvider` and `getLogsSampleDataProvider`, so there is no code duplication. 

I added tests for this new method for both data sources that are `DataSourceWithSupplementaryQueriesSupport`  - Elasticsearch and Loki. 

I also considered other options such as:

- reusing `executedQueryString` that is in Logs data frames, but `splitOpen` requires full DataQuery so this wouldn't work
- saving `executedQueries` in redux, but currently we use data source directly for all of the smaller interactions such as  adhoc filtering, log filtering, log context, ... . So I'd decided to use the same pattern for these smaller interactions and in case we decide to change this pattern, we can refactor this as well. 

How to test:

1. Start loki data source `make devenv sources=loki`
2. Open Explore and run metric query - e.g. `rate({place="moon"}[5m)`
4. In logs sample, click on `Open logs in split view`
5. Split view with log query should be open
